title: "Assignment-3 Rmd report"
author: "Joseph Kinyanjui"
date:
output:
    rmarkdown::html_document:
        toc: true
theme: cosmo
---
SOURCE URL: https://github.com/Josephgatonye/Assignment-3/tree/master

# **Determining the limits of BLAST assignment**
# Introduction
This work will apply BLAST(Basic Local ALignment Search Tool) to run large scale analyses in rstudio. 
# Installing needed library packages
There are some library packages that will be installed first to enable smooth working. These packages are used to load in data to R
```{r,libs}
library("seqinr")
library("R.utils")
library("rBLAST")
```
or by using 
```install.packages(c("dplyr", "purrr", "tidyr")

# Question 1
# Downloading the set of E.coli gene DNA sequence 
```{r,chunk1}
download.file("ftp://ftp.ensemblgenomes.org/pub/bacteria/release-42/fasta/bacteria_0_collection/escherichia_coli_str_k_12_substr_mg1655/cds/Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.cds.all.fa.gz",destfile = "E.coli.fa.gz")
gunzip("E.coli.fa.gz")
makeblastdb("E.coli.fa",dbtype = "nucl","-parse_seqids") #create blast and gives the numbers of sequences present in the E.coli set
```
#Question 2
```{r,chunk2}
# Downloading the sample fasta sequences from github 
download.file("https://raw.githubusercontent.com/markziemann/SLE712_files/master/bioinfo_asst3_part2_files/sample.fa",destfile = "sample.fa")
# Reading file into R
This has to be done just within R. 
sample <- read.fasta("sample.fa")
myseq <- sample[[27]] #extracting allocated sequence 
myseq
# Calculating GC contents
str(myseq)
seqinr::getLength(myseq)# calculating the length (in bp)of GC bases
seqinr::GC(myseq) # calculating the proportion of GC bases
```
#Question 3
A function that runs blast searches was sourced from unix shell tools.

# Blast search for E.coli genome with the highest similarity to my sample sequence
Here we are searching the sample sequence against the contents of the set of E.coli gene DNA sequence.
This can give clues to the origin of the sample sequence.  
```{r,blast1}
#myblastn is a wrapper for the unix blastn tool
#the input is an R sequence and the output is tabular object
download.file("https://raw.githubusercontent.com/markziemann/SLE712_files/master/bioinfo_asst3_part2_files/mutblast_functions.R",destfile = "mutblast.R")
source("mutblast.R")
#test the function
res <- myblastn_tab(myseq = myseq,db = "E.coli.fa")
# Check out the blast results
res
# Gene ID :fpr b3924
# Description:ferredoxin-NADP reductase; flavodoxin reductase
#About gene: This gene has 1 transcript
#Gene type: Protein coding
str(res)
head(res)
hits <- as.character(res$sseqid[1:3])# It is crucial to store the sseqid
hits# sseqid will be used to determine whether BLAST matches the gene of origin
```
#Question 4
Creating a mutated copy with multiple substitutions
```{r,testblast}
# writing the blast index
write.fasta(myseq,names="myseq",file.out ="myseq.fa")
makeblastdb(file = "myseq.fa",dbtype = "nucl")

#test with 20 mutations
myseq_mut <-mutator (myseq,50)
res1 <- myblastn_tab(myseq = myseq_mut,db = "myseq.fa")
res1
myseq_mut <-mutator (myseq,100)
res1 <- myblastn_tab(myseq = myseq_mut,db = "myseq.fa")
res1
myseq_mut <-mutator (myseq,150)
res1 <- myblastn_tab(myseq = myseq_mut,db = "myseq.fa")
res1

myseq_mut <-mutator (myseq,175)
res1 <- myblastn_tab(myseq = myseq_mut,db = "myseq.fa")
res1
myseq_mut <-mutator (myseq,200)
res1 <- myblastn_tab(myseq = myseq_mut,db = "myseq.fa")
res1
```




