---
title: "Bioinformatics Assignment 3: Importing files, data wrangling, mathematical operations, plots and saving code on GitHub Report"
author: "Joseph Kinyanjui and Faith Kosgei"
date: "12/06/2020"
output:
    rmarkdown::html_document:
        toc: true
theme: cosmo
---

Source: https://github.com/Josephgatonye/Assignment-3/tree/master

# Gene expression

<<<<<<< HEAD
## Question 1 
The `download.file()`  command is used to fetch the dataset. The `read.table()` command  is used to read the tsv file into R and the data frame is then saved as an object for easy computation. The `row.names=1` commmand is used to make the gene accession numbers the row names. It is also crucial to set 'header=TRUE' and `stringAsfactors= FALSE` so that the data is formatted properly. The `head ()`command is used to show the first six lines of the data frame and the `str ()` command checks that the data structure is recognized as numeric. 
=======
## Question 1 - downloading the file, reading into R, making the gene accession numbers the row names, and showing a table of values for the first 6 genes
- Here the 'download.file()' command is used to fetch the dataset and 'read.table()' command  is used to read the tsv file into R from an external source.
- The data frame is then saved as object "x" for easy computation using the '<-' command.
- The 'row.names=1' commmand is used to make the gene accession numbers the row names.
- It is also crucial to set 'header=TRUE' and 'stringAsfactors= FALSE' so that the data is formatted properly. - The 'head ()'command is used to show the first six lines of the data frame and the 'str ()' command ensures that the data structure is recognized as numeric. 
>>>>>>> d176c13b86a784d4526949c8320396630edcdb07
```{r,chunk1,echo=TRUE}
#Downloading the file and saving it as "gene_expression.tsv"
download.file("https://raw.githubusercontent.com/markziemann/SLE712_files/master/bioinfo_asst3_part1_files/gene_expression.tsv",destfile = "gene_expression.tsv") 

#saving the file as an object x
x <- read.table("gene_expression.tsv") 

#cheking the data structure
head(x) 
str(x) 

#Reading into R making sure the 1st columns is set as the rownames, headers are recognized an only first 6 genes are shown
x <- read.table("gene_expression.tsv",header=TRUE,row.names= 1, stringsAsFactors = FALSE)

# Showing showing a table of values for the first 6 genes
head(x)
str(x)
```
## Question 2 
We are creating a new column which is the mean of the previous columns using the `rowMeans(x)` command, labelling the new column as `ColumnMeans` and adding it permanently into the data frame x using the `$` command. The table of values for the first six genes is showed using the 'head ()' command 
```{r,chunk2,echo=TRUE}
# Making a new column ("ColumnMeans") which is the mean of the other columns
x$ColumnMeans<- rowMeans(x)

# Showing a table of values for the first six genes
head(x)
```
## Question 3 
We already have a column with means of genes saved and it is ordered using `order(-)` command to show decreasing order of means of genes and saved as object `x_ordered`. The `list ()`command is used to show the data as a list and square brackets command used to specify a range of the first ten genes in the new data frame.
```{r,chunk3,echo=TRUE}
# Listing the 10 genes with the highest mean expression
x_ordered <- x[order(-x$ColumnMeans),]
list(x_ordered [1:10,])
```
## Question 4 
Here we will subset the data frame using the `subset()`command and extract the column which contains means of the genes which is less than 10. The `nrow()` command gives the number of rows in the data frame.

```{r,chunk4,echo=TRUE}
# Subseting the number of genes with a mean less than 10
no_genes <- subset(x, ColumnMeans <10)
nrow(no_genes) # provides the number of genes as 43124
```
## Question 5 
The first step is to create a png file using the png() command and add the histogram plot created with the hist () command. The second step is to import the file into the report.

```{r,chunk5,echo=TRUE}
png("gene_expression.png") 
hist(x$ColumnMeans,main = "Histogram of mean values",xlab ="column means")
dev.off()
```
![alt text](gene_expression.png "Gene_Expression Plot") 

# Growth data

<<<<<<< HEAD
## Question 6 
The `growth_data.csv` file is downloaded and read into R using the `read.csv ()` command. The `colnames ()`command is used to extract the names of the columns in the data frame.
=======
# Question 6 
# Importing the csv file into an R object
The backslash is used within the path name to be copied to avoid any eror while copying.
>>>>>>> d176c13b86a784d4526949c8320396630edcdb07
```{r,chunk6,echo=TRUE}
download.file("https://raw.githubusercontent.com/markziemann/SLE712_files/master/bioinfo_asst3_part1_files/growth_data.csv",destfile = "growth_data.csv")
# Reading csv file into R
g<- read.csv("growth_data.csv")
Letting R know that this is comma separeted values
g<- read.csv("file.choose(), header=T, Sep= ",")
head(g)
str(x)
# Finding the column names
colnames(g)
```

## Question 7
We begin by first subsetting the data frame into the two control sites (north east and southwest) and then extracting the columns with the data at the start and the end of the study. Mean is calculated using the `mean ()`command and standard deviation calculated using the `sd()` command.
```{r,chunk71,echo=TRUE}
ne <- subset(g,Site== "northeast") #subsetting site upto northeast and saving as ne
#checking structure
head(ne)
str(ne)
# getting the mean of circumfrence at the start
mean(ne$Circumf_2004_cm, ) 
# getting the mean of circumfrence at the end
mean(ne$Circumf_2019_cm)
#getting standard deviation of  circumfrence at the start
sd(ne$Circumf_2004_cm)
#getting standard deviation of  circumfrence at the end
sd(ne$Circumf_2019_cm)
```

```{r,chunk72,echo=TRUE}
#subsetting site upto southwest and saving as sw
sw <- subset(g,Site== "southwest")

# checking the data structure
head(sw)
str(sw)

# getting the mean of circumfrence at the start
mean(sw$Circumf_2004_cm)

# getting the mean of circumfrence at the end
mean(sw$Circumf_2019_cm)

#getting standard deviation of  circumfrence at the start
sd(sw$Circumf_2004_cm)

#getting standard deviation of  circumfrence at the end
sd(sw$Circumf_2019_cm)
```
Box plots divides the data into three parts and therefore the data distribution can be easily visualised.

## Question 8
The `boxpot()` command used to make a boxplot of tree circumference at the start and end of the study at both sites.
```{r,chunk8-1,echo=TRUE}
# box plot of tree circumference at the start and end of the study at north west 
boxplot(ne$Circumf_2004_cm,ne$Circumf_2019_cm)
```

```{r,chunk8-2,echo=TRUE}
 # box plot of tree circumference at the start and end of the study at south east
boxplot(sw$Circumf_2004_cm,sw$Circumf_2019_cm)
```
The boxplots show non-equal variances.
## Question 9
The growth difference over the past 10 years at each site is calculated by subtraction using the recent data (2019 and 2009) and the result saved as an R object. The `mean()`command is then applied to the saved object to calculate the mean.
North east site
```{r,chunk9-1,echo=TRUE}
#getting the growth difference over 10 years(2009$2019) @ North east
growthdif1<- (ne$Circumf_2019_cm - ne$Circumf_2009_cm)
# getting the mean of growth difference
mean(growthdif1)
```
South east site
```{r,chunk9-2,echo=TRUE}
#getting the growth difference over 10 years(2009$2019) @ south west
growthdif2<- (sw$Circumf_2019_cm - sw$Circumf_2009_cm)
# getting the mean of growth difference
mean(growthdif2)
```

## Question 10
A t test shows the significance of the differences between two groups which can be used to find out if the differences could haver occured by chance (Statistics How To 2020). The Wilcoxon rank sum test checks if the distributions of observations acquired between two separate groups are significantly different from each other (Salkind, N.J 2010). The t.test () and wilcox.test () commands will be used to estimate the p-value that the 10 year growth is different at the two sites.

```{r,chunk10-1,echo=TRUE}
t.test(growthdif1,growthdif2)
```
```{r,chunk10-2,echo=TRUE}
wilcox.test(growthdif1,growthdif2)
```
<<<<<<< HEAD
P- values show that there is a huge difference between growth in the sites

# References

Statistics How To 2020, 
T Test (Studentâ€™s T-Test): Definition and Examples, retrieved June 2 2020, <https://www.statisticshowto.com/probability-and-statistics/t-test/>.

Salkind, N.J 2010, Encyclopedia of research design, vol. 0, SAGE Publications, Inc., Thousand Oaks, CA, [Accessed 12 June 2020], doi: 10.4135/9781412961288.



=======
t.test is used to test difference between the means of two groups)# For purpose of hypothesis testing in statistics
calculating t.test requires difference between the mean values from each data  # (p-value <1)                                                                                                                                       
wilcoxon test is  the rank sum test that compares two paired groups to determine if two or more sets of pairs differ from one another in terms of statistically'
>>>>>>> d176c13b86a784d4526949c8320396630edcdb07
